core/shared/src/main/scala/org/typelevel/log4cats/SamStructuredLogger.scala

```diff
@@ -0,0 +1,326 @@
+/*
+ * Copyright 2018 Typelevel
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.typelevel.log4cats
+
+import cats.*
+import cats.data.{EitherT, Kleisli, OptionT}
+import cats.Show.Shown
+
+/**
+ * A SAM-based StructuredLogger that extends Logger and provides structured logging capabilities.
+ * This implementation uses the new SAM LoggerKernel design for better performance and middleware
+ * compatibility.
+ */
+trait SamStructuredLogger[F[_]] extends Logger[F] {
+  protected def kernel: LoggerKernel[F, String]
+
+  def trace(ctx: Map[String, String])(msg: => String): F[Unit] = {
+    val builder = (b: Log.Builder[String]) => {
+      var current = b.withMessage(msg)
+      ctx.foreach { case (k, v) => current = current.withContext[String](k)(v) }
+      current
+    }
+    kernel.logTrace(builder)
```

Comment:
Might be worth exercising the methods on `Log.Builder` to see if they pull their weight. 
For example, this can be simplified to:
```suggestion
    kernel.logTrace(_.withMessage(msg).withContextMap(ctx))
```